File: app/src/main/java/com/example/focusstartsecondpart/features/events/data/EventsDataSourceImpl.java
package com.example.focusstartsecondpart.features.events.data;

import com.example.focusstartsecondpart.App.App;
import com.example.focusstartsecondpart.features.events.data.network.EventsLoader;
import com.example.focusstartsecondpart.features.events.domain.model.Event;

import java.util.List;

import io.reactivex.Single;

public class EventsDataSourceImpl implements EventsDataSource {

    private EventsLoader eventsLoader;

    public EventsDataSourceImpl(EventsLoader eventsLoader) {
        this.eventsLoader = eventsLoader;
    }

    @Override
    public Single<List<Event>> loadEvents() {
        return loadEventsFromNet()
                .onErrorResumeNext(throwable -> loadEventsFromDatabase());
    }

    private Single<List<Event>> loadEventsFromNet() {
        return eventsLoader.loadEvents().doOnSuccess(events -> {
            App.getDataBase().getDatabaseDao().insertAllEvents(events);
        });
    }
    private Single<List<Event>> loadEventsFromDatabase() {
         return App.getDataBase().getDatabaseDao().getAllEvents();
    }
}

File: app/src/main/java/com/example/focusstartsecondpart/features/events/data/EventsDataSourceImpl.kt
package com.example.focusstartsecondpart.features.events.data

import com.example.focusstartsecondpart.App.App
import com.example.focusstartsecondpart.features.events.data.network.EventsLoader
import com.example.focusstartsecondpart.features.events.domain.model.Event

import io.reactivex.Single

class EventsDataSourceImpl(private val eventsLoader: EventsLoader) : EventsDataSource {

    override fun loadEvents(): Single<List<Event>> {
        return loadEventsFromNet()
                .onErrorResumeNext {loadEventsFromDatabase() }
    }

    private fun loadEventsFromNet(): Single<List<Event>> {
        return eventsLoader.loadEvents()
                .doOnSuccess { events -> App.getDataBase().databaseDao.insertAllEvents(events) }
    }

    private fun loadEventsFromDatabase(): Single<List<Event>> {
        return App.getDataBase().databaseDao.allEvents
    }
}
